certora/specs/TimelockController.spec

```diff
@@ -0,0 +1,278 @@
+└─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┘
+*/
+
+definition DONE_TIMESTAMP()      returns uint256 = 1;
+definition UNSET()               returns uint8   = 0;
+definition PENDING()             returns uint8   = 1;
+definition DONE()                returns uint8   = 2;
+
+definition isUnset(bytes32 id)   returns bool    = !isOperation(id);
+definition isPending(bytes32 id) returns bool    = isOperationPending(id);
+definition isDone(bytes32 id)    returns bool    = isOperationDone(id);
+definition state(bytes32 id)     returns uint8   = isUnset(id) ? UNSET() : isPending(id) ? PENDING() : DONE();
+
+/*
+┌─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┐
+│ Invariants: consistency of accessors                                                                                │
+└─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┘
+*/
+invariant isOperationCheck(bytes32 id)
+    isOperation(id) <=> getTimestamp(id) > 0
+    filtered { f -> !f.isView }
+
+invariant isOperationPendingCheck(bytes32 id)
+    isOperationPending(id) <=> getTimestamp(id) > DONE_TIMESTAMP()
+    filtered { f -> !f.isView }
+
+invariant isOperationDoneCheck(bytes32 id)
+    isOperationDone(id) <=> getTimestamp(id) == DONE_TIMESTAMP()
+    filtered { f -> !f.isView }
+
+invariant isOperationReadyCheck(env e, bytes32 id)
+    isOperationReady(e, id) <=> (isOperationPending(id) && getTimestamp(id) <= e.block.timestamp)
+    filtered { f -> !f.isView }
+
+/*
+┌─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┐
+│ Invariant: a proposal id is either unset, pending or done                                                           │
+└─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┘
+*/
+invariant stateConsistency(bytes32 id, env e)
+    (isUnset(id)   <=> state(id) == UNSET()              ) &&
+    (isPending(id) <=> state(id) == PENDING()            ) &&
+    (isDone(id)    <=> state(id) == DONE()               ) &&
+    (isUnset(id)   <=> (!isPending(id) && !isDone(id)   )) &&
+    (isPending(id) <=> (!isUnset(id)   && !isDone(id)   )) &&
+    (isDone(id)    <=> (!isUnset(id)   && !isPending(id))) &&
+    isOperationReady(e, id) => isPending(id)
+    filtered { f -> !f.isView }
+
+/*
+┌─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┐
+│ Rule: state transition rules                                                                                        │
+└─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┘
+*/
+rule stateTransition(bytes32 id, env e, method f, calldataarg args) {
```

Comment:
That is a possibility. I considered many ways of writing it. If you feel like the current syntax is not clear enough and you have something to propose, fell free!

Reply:
The current syntax is very clear and I like it, and as far as I can tell it is exhaustive. But the tool isn't helping us to check that we are exhaustively checking all possible before and after state combinations.

I was thinking that if we write something like this, we can be sure because of all the final `else` statements that there are no combinations that we missed. I definitely prefer the current spec aesthetically though.

```
if (stateBefore == UNSET()) {
    if (stateAfter == PENDING()) {
        // UNSET → PENDING: schedule or scheduleBatch
        assert (
            f.selector == schedule(address, uint256, bytes, bytes32, bytes32, uint256).selector ||
            f.selector == scheduleBatch(address[], uint256[], bytes[], bytes32, bytes32, uint256).selector
        );
    } else {
        assert stateAfter == UNSET();
    }
} else if (stateBefore == PENDING()) {
    if (stateAfter == UNSET()) {
        // PENDING → UNSET: cancel
        assert stateBefore == PENDING() && stateAfter == UNSET() => (
            f.selector == cancel(bytes32).selector
        );
    } else if (stateAfter == DONE()) {
        // PENDING → DONE: execute or executeBatch
        assert stateBefore == PENDING() && stateAfter == DONE() => (
            f.selector == execute(address, uint256, bytes, bytes32, bytes32).selector ||
            f.selector == executeBatch(address[], uint256[], bytes[], bytes32, bytes32).selector
        );
    } else {
        assert stateAfter == PENDING();
    }
} else if (stateBefore == DONE()) {
    // DONE is final
    assert stateAfter == DONE();
} else {
    assert false;
}
```

Reply:
If we had `switch` it would be so clean